docs/tokencache.md

```diff
@@ -2,16 +2,350 @@
 
 ## Introduction
 
-> PENDING
+The token cache implementation for MSAL, ADAL and Android broker is found in the Android Common Library.  
 
-## Class Diagram
+## References 
 
-> PENDING
+- Universal Cache Schema
 
-## Interaction Diagram
+## Universal cache and protocol response handling
 
-> PENDING
+When looking at the token cache there are really 2 levels to it.  The lower level is based on the universal cache schema referenced above.  The higher level deals with the result of a particular protocol request and mapping that to the universal schema.  
+
+So in effect we get an oAuth 2 Token Response, which in the case of MSAL includes an ID Token and an Access Token and we translate:
+
+ID Token & Client_Info -> Account & Tenant Profile
+Access Token -> Credential
+Refresh Token -> Credential
+
+## Universal Schema at a high level / OAuth v2 Mapping
+
+- Accounts
+  - ID Token of account home tenant
+- Tenant Profiles - The representation of 1 account in 1 or more tenants (external account guest/member scenarios)
+  - ID Token of the account home tenant
+  - ID Token from other tenants that the account is a member of
+- Credentials
+  - Id Token
+  - Access Tokens
+  - Refresh Tokens
+  - etc....
+
+### Relationships
+
+1 Account has 0 or more credentials
+1 Credential belongs to 1 account
+
+## OAuth 2 Token Cache
+
+This section describes the base class for the higher level of the TokenCache.  The level responsible for translated an oAuth response into universal cache schema artifacts.
+
+There are different implementations of the cache based on the library using it and the scenarios it needs to support.
+
+> NOTE: I'm going to ignore the ADALOAuth2TokenCache in this document. It deals with interacting with the legacy ADAL token cache.
+
+### Abstract base class
+
+The abstract base class for all caches is the OAuth2TokenCache.  The physical schema of the cache in terms of account and credential records is decoupled from the protocol used to acquire the account and credential; however the cache interface knows about oAuth2 token responses and how to convert those to accounts and crentials and persist them to storage.
```

Comment:
Accurate -- _ideally_ the `OAuth2TokenCache` should delegate to the strategy (or an object provided by the strategy) to produce the account/credential types... this wasn't implemented this way, but with the benefit of hindsight, this would be a future "nice-to-have"